compiler/src/iree/compiler/Codegen/TransformDialectExtensions/TransformDialectExtensions.cpp

```diff
@@ -41,6 +46,315 @@ void mlir::iree_compiler::registerLinalgTransformDialectExtension(
+  // Step 2. Get values defined above and separate them between non-tensors,
+  // tensors and introduce appropriate tensor dims.
+  // Tensors that have already been recorded as resultTensorOperands are
+  // omitted to avoid duplications.
+  llvm::SetVector<Value> valuesDefinedAbove;
+  mlir::getUsedValuesDefinedAbove(foreachThreadOp.getRegion(),
+                                  valuesDefinedAbove);
+  assert(!valuesDefinedAbove.empty() && "used values defined above is empty");
+
+  SmallVector<Value> nonTensorOperands, tensorOperands, tensorDynamicDims;
+  for (Value v : valuesDefinedAbove) {
+    auto tensorType = v.getType().dyn_cast<RankedTensorType>();
+    if (!tensorType) {
+      nonTensorOperands.push_back(v);
+      continue;
+    }
+    if (resultTensorOperands.contains(v)) continue;
+    tensorOperands.push_back(v);
+    for (int64_t dim : getIndicesOfDynamicDims(tensorType))
+      tensorDynamicDims.push_back(rewriter.create<tensor::DimOp>(loc, v, dim));
+  }
+
+  // Step 3. Compute workgroupCount.
+  // TODO: this should actually be workLoad information.
+  // Also this will require special handling during codegen to set up the
+  // HAL::EntryPointOp correctly (atm only 1x1x1 is supported).
+  SmallVector<Value> workgroupCount{foreachThreadOp.getNumThreads()};
+  Value one = rewriter.create<arith::ConstantIndexOp>(loc, 1);
+  workgroupCount.resize(3, one);
+
+  // Step 4. Create ordered vectors of operands to pass to the builder and
+  // build the dispatchOp.
+  SmallVector<Value> nonDimOperands;
+  llvm::append_range(nonDimOperands, nonTensorOperands);
+  llvm::append_range(nonDimOperands, tensorOperands);
+  llvm::append_range(nonDimOperands, resultTensorOperands);
+  // scf::ForeachThreadOp tensors inserted into are tied to results and
+  // translate to the tied operands of the dispatch.
+  int64_t sizeNonTensors = nonTensorOperands.size();
+  int64_t sizeNonResultTensors = tensorOperands.size();
+  int64_t sizeResultTensors = resultTensorOperands.size();
+  auto tiedOperandsSequence = llvm::seq<int64_t>(
+      sizeNonTensors + sizeNonResultTensors,
+      sizeNonTensors + sizeNonResultTensors + sizeResultTensors);
+  // Separate out tensorOperands and tensorDynamicDims for RAUWIf purposes.
+  SmallVector<Value> allTensorOperands = tensorOperands;
+  llvm::append_range(allTensorOperands, resultTensorOperands);
+  SmallVector<Value> allTensorDynamicDims = tensorDynamicDims;
+  llvm::append_range(allTensorDynamicDims, resultTensorsDynamicDims);
+  // clang-format off
+  auto dispatchOp = rewriter.create<Flow::DispatchWorkgroupsOp>(
+      loc,
+      /*workgroupCount=*/workgroupCount,
+      /*resultTypes=*/foreachThreadOp.getResultTypes(),
+      /*resultDims=*/resultTensorsDynamicDims.getArrayRef(),
+      /*operands=*/nonDimOperands,
+      /*operandDims=*/allTensorDynamicDims,
+      /*tiedOperands=*/llvm::to_vector<4>(tiedOperandsSequence));
+  // clang-format on
+
+  // Step 5. Fixup dispatchOp bbArgs and terminator.
```

Comment:
Yeah I saw that the helper function that infers stuff is clever and allows arbitrary mixing of tensors and dims as long as the dims come after the tensors.
In the op itself, it would be nice to have some verification and at least document an expected way we would want to read them.